File: linux/gantry/panelwindow.py
"""One auxiliary panel in a window of its own, centred on the screen, and the header every panel wears.

The panels used to be overlays inside the fleet window: a dimmed layer with the card floating in the
middle of it. That cost more than it bought. A panel could never be larger than its host, it dimmed
the cards the user had just come to read, and the fleet window had to be kept awake by hand so it did
not hide from under the thing it had opened.

Every panel window follows the `panelWindow` contract in design/gantry-card-layout.impl.json, the
same one macOS PanelWindowController and Windows PanelWindow follow: a header bar carrying the fleet
panel's own identity, GANTRY · name, at the leading edge, the panel's own controls at the trailing
edge, the system close button, and a window title of "Gantry · name". The panels no longer draw a
title or a close button of their own.
"""
from __future__ import annotations

from typing import Any, Callable, Iterable

from gi.repository import Gtk, Gdk  # type: ignore


def window_title(name: str) -> str:
    """ "Gantry · Spoolbase". A middle dot, as in the fleet header. """
    return f"Gantry · {name}"


def panel_header(window: Gtk.Window, name: str, accessories: Iterable[Gtk.Widget] = ()) -> Gtk.HeaderBar:
    """Installs the shared header on `window` and returns it.

    Works on a Gtk.Dialog as well as a Gtk.Window: a dialog keeps its own action buttons at the bottom
    and only its title bar is replaced.
    """
    header = Gtk.HeaderBar()
    header.set_show_close_button(True)
    # Leading, not the header bar's centred title: the wordmark sits where it sits in the fleet
    # header, and where macOS puts it beside the traffic lights. An empty custom title keeps GTK from
    # drawing the plain title text in the middle as well.
    header.set_custom_title(Gtk.Box())
    identity = Gtk.Box(spacing=7)
    wordmark = Gtk.Label(label="GANTRY")
    wordmark.get_style_context().add_class("panel-wordmark")
    dot = Gtk.Label(label="·")
    dot.get_style_context().add_class("panel-dot")
    title = Gtk.Label(label=name, xalign=0)
    title.set_ellipsize(3)  # Pango.EllipsizeMode.END, without importing Pango for one constant
    title.get_style_context().add_class("panel-name")
    for widget in (wordmark, dot, title):
        identity.pack_start(widget, False, False, 0)
    header.pack_start(identity)
    # pack_end fills from the right edge inwards, so reverse to keep the caller's left-to-right order.
    for widget in reversed(list(accessories)):
        header.pack_end(widget)
    header.show_all()
    window.set_titlebar(header)
    # After set_titlebar, not before: installing a header bar takes the window's title over, and a
    # title set earlier reads back empty (measured: None). The custom title keeps it from being drawn.
    header.set_title(window_title(name))
    window.set_title(window_title(name))
    return header


class PanelWindow(Gtk.Window):
    """A bordered, resizable window holding one panel widget, centred on the screen."""

    def __init__(self, app: Any, content: Gtk.Widget, name: str,
                 width: int = 470, height: int = 560,
                 cleanup: Callable[[], None] | None = None,
                 accessories: Iterable[Gtk.Widget] = ()) -> None:
        super().__init__()
        self.app = app
        self._cleanup = cleanup
        self.set_transient_for(app.window)
        # Centred on the screen, not on the parent: a panel centred on the fleet window opens on top
        # of the cards, which is the thing the overlay did wrong.
        self.set_position(Gtk.WindowPosition.CENTER)
        self.set_default_size(width, height)
        self.set_size_request(min(360, width), min(240, height))
        panel_header(self, name, accessories)
        # The panel drew its own centred card because it had a whole window's worth of dimmed space
        # around it. In a window of this size it is the content, so it fills it.
        content.set_halign(Gtk.Align.FILL)
        content.set_valign(Gtk.Align.FILL)
        content.set_size_request(-1, -1)
        scroll = Gtk.ScrolledWindow()
        scroll.set_policy(Gtk.PolicyType.AUTOMATIC, Gtk.PolicyType.AUTOMATIC)
        scroll.add(content)
        self.add(scroll)
        self.connect("key-press-event", self._on_key)
        self.connect("destroy", self._on_destroy)
        # The fleet window stays on screen for as long as this one does.
        app.window.hold_fleet_panel(self)

    def _on_key(self, _widget: Gtk.Widget, event: Gdk.EventKey) -> bool:
        if event.keyval == Gdk.KEY_Escape:
            self.destroy()
            return True
        return False

    def _on_destroy(self, *_args: object) -> None:
        cleanup, self._cleanup = self._cleanup, None
        if cleanup:
            cleanup()

    def present_centered(self) -> None:
        self.show_all()
        self.present()

File: windows/Gantry.Windows/UI/PanelWindow.cs
using System.Windows;
using System.Windows.Controls;
using System.Windows.Input;
using System.Windows.Media;

namespace Gantry.UI;

/// <summary>One auxiliary panel in a window of its own, centred on the screen.
///
/// Slot assignment and maintenance were dimmed overlays inside the fleet panel. That cost more than
/// it bought: the panel could never be larger than its host, it blurred and disabled the cards the
/// user had just come to read, and the fleet panel had to be told by hand not to hide from under the
/// thing it had opened. Mirrors macOS PanelWindowController and Linux panelwindow.PanelWindow.
///
/// Every panel window follows the panelWindow contract in design/gantry-card-layout.impl.json: the
/// fleet panel's own identity, GANTRY · name, in a header strip under the system title bar, the
/// panel's own controls at its trailing end, and "Gantry · name" as the window title. The panels no
/// longer draw a title or a close button of their own.</summary>
internal sealed class PanelWindow : Window
{
    private readonly Action? _cleanup;

    /// <summary>"Gantry · Spoolbase". Title bar, taskbar and Alt+Tab all show this string; the header
    /// draws the same two parts with the wordmark. A middle dot, as in the fleet header.</summary>
    public static string WindowTitle(string name) => $"Gantry · {name}";

    /// <summary>The shared header strip: GANTRY · name on the leading edge, accessories trailing, a
    /// hairline under it.</summary>
    public static FrameworkElement Header(string name, IReadOnlyList<UIElement>? accessories = null)
    {
        var identity = new StackPanel { Orientation = Orientation.Horizontal, VerticalAlignment = VerticalAlignment.Center };
        // The fleet header's wordmark, a size down: this one shares the window with a title bar.
        identity.Children.Add(new TextBlock
        {
            Text = "GANTRY", FontSize = 15, FontWeight = FontWeights.Black,
            FontFamily = new FontFamily("Segoe UI Black, Segoe UI"),
            Foreground = GTheme.Brush(GTheme.Text), VerticalAlignment = VerticalAlignment.Center,
        });
        identity.Children.Add(new TextBlock
        {
            Text = "·", FontSize = 13, FontWeight = FontWeights.SemiBold, Margin = new Thickness(7, 0, 7, 0),
            Foreground = GTheme.Brush(GTheme.Secondary), VerticalAlignment = VerticalAlignment.Center,
        });
        identity.Children.Add(new TextBlock
        {
            Text = name, FontSize = 14, FontWeight = FontWeights.SemiBold, TextTrimming = TextTrimming.CharacterEllipsis,
            Foreground = GTheme.Brush(GTheme.Text), VerticalAlignment = VerticalAlignment.Center,
        });
        var trailing = new StackPanel { Orientation = Orientation.Horizontal, VerticalAlignment = VerticalAlignment.Center };
        if (accessories is not null)
            foreach (var accessory in accessories) trailing.Children.Add(accessory);
        var row = new Grid();
        row.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        row.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
        row.Children.Add(identity);
        Grid.SetColumn(trailing, 1);
        row.Children.Add(trailing);
        return new Border
        {
            Padding = new Thickness(16, 9, 12, 9), MinHeight = 44,
            BorderBrush = GTheme.Brush(GTheme.Line), BorderThickness = new Thickness(0, 0, 0, 1),
            Child = row,
        };
    }

    /// <summary>Puts the shared header above <paramref name="body"/> in <paramref name="window"/> and
    /// sets its title. For panels that were already Windows of their own (diagnostics, statistics,
    /// Spoolbase), so they wear exactly the frame a PanelWindow does.</summary>
    public static void Wrap(Window window, string name, FrameworkElement body, IReadOnlyList<UIElement>? accessories = null)
    {
        window.Title = WindowTitle(name);
        var header = Header(name, accessories);
        DockPanel.SetDock(header, Dock.Top);
        window.Content = new DockPanel { LastChildFill = true, Children = { header, body } };
    }

    /// <param name="scrolls">true only for content with no scroll viewer of its own. Wrapping content
    /// that already scrolls in a second scroll viewer hands the inner one infinite height, so it
    /// never scrolls and the outer one drags the panel's header off the top instead.</param>
    public PanelWindow(Window owner, FrameworkElement content, string name,
                       double width = 470, double height = 650, bool scrolls = false,
                       Action? cleanup = null, IReadOnlyList<UIElement>? accessories = null)
    {
        _cleanup = cleanup;
        Width = width;
        Height = height;
        MinWidth = Math.Min(360, width);
        MinHeight = Math.Min(240, height);
        // Centred on the screen, not on the owner: a panel centred on the fleet panel opens straight
        // on top of the cards, which is the thing the overlay did wrong.
        WindowStartupLocation = WindowStartupLocation.CenterScreen;
        ShowInTaskbar = false;
        // Owned by the fleet panel for two reasons: it always draws above it, and the panel's
        // Deactivated handler walks OwnedWindows and refuses to hide while any of them is visible.
        // Owner is only legal once the owner itself has been shown.
        if (owner.IsLoaded) Owner = owner;

        // The panel drew its own rounded card because it floated in a whole window's worth of dimmed
        // space. Here it is the window's content, so it fills it and the window carries the surface.
        content.Margin = new Thickness(0);
        content.HorizontalAlignment = HorizontalAlignment.Stretch;
        content.VerticalAlignment = VerticalAlignment.Stretch;
        // An explicit Width, Height or Max* beats Stretch in WPF, and both overlay panels pinned one
        // (maintenance 470x560, slot assignment its longest name and 440 tall). In a window they fill.
        content.Width = double.NaN;
        content.Height = double.NaN;
        content.MaxWidth = double.PositiveInfinity;
        content.MaxHeight = double.PositiveInfinity;
        if (content is Border card)
        {
            card.CornerRadius = new CornerRadius(0);
            card.BorderThickness = new Thickness(0);
        }
        FrameworkElement body;
        if (scrolls)
        {
            body = new ScrollViewer
            {
                Content = content,
                VerticalScrollBarVisibility = ScrollBarVisibility.Auto,
                // Off on purpose, the same as the overlay it replaces: content slightly wider than
                // the window should lay out narrower, not grow a scrollbar across the bottom.
                HorizontalScrollBarVisibility = ScrollBarVisibility.Disabled,
                Background = Brushes.Transparent,
                Padding = new Thickness(0),
            };
        }
        else
        {
            body = content;
        }
        Wrap(this, name, body, accessories);
        // After Content, not before: ApplyWindowTheme also walks the tree to theme fields and
        // buttons, and with Content still null there is nothing to walk.
        GTheme.ApplyWindowTheme(this);

        PreviewKeyDown += (_, e) =>
        {
            if (e.Key != Key.Escape) return;
            e.Handled = true;
            Close();
        };
        Closed += (_, _) => _cleanup?.Invoke();
    }

    /// <summary>Shows the panel and brings it forward.</summary>
    public static PanelWindow Present(Window owner, FrameworkElement content, string name,
                                      double width = 470, double height = 650, bool scrolls = false,
                                      Action? cleanup = null, IReadOnlyList<UIElement>? accessories = null)
    {
        var window = new PanelWindow(owner, content, name, width, height, scrolls, cleanup, accessories);
        window.Show();
        window.Activate();
        return window;
    }
}
